### ISSUE TITLE
Regex exception when escaping [ in a character class

### ISSUE THREAD

**Issue by bertbarabas** [guillep/pharo-core#8393]
asRegex fails with the exception:
RegexSyntaxError: no terminating "]" 
when compiling a character class with an escaped [ character.

**To Reproduce**
1. Go to a workspace
2. doit for the following code: `'[\[]'asRegex`
3. See error

**Expected behavior**
There should be no exception and the the resulting regular exception should be able to match a single [ character.

**Version information:**
 - OS: [Windows 10]
 - Version: [Pharo 8.0 - 64bit (stable).image]

**Expected development cost**
As you know nothing happens magically, so do you have an idea of the development effort required? 
In addition, are you willing to help because after all Pharo is yours too?

It doesn't seem like it should be complex, hopefully just a small bug. I'm happy to help although I'm not familiar with the details of the regex package.

**Comment by merkste** [comment]
According to the specification:

1. '[\]]' asRegex is not supposed to work as \] is not allowed in character sets. Use '[]]' instead.
2. '[\[]' asRegex is also not supposed to work for the same reason.

However, some related bugs prevent the parser vom recognizing '[[]' as a valid RegEx. I fixed these in RegEx11 (v 1.4.7) on which Pharo's RegEx are based. I am planning to bring recent fixes for RegEx11 to Pharo soon but didn't have the time to do it yet. In particular:

1. Fix: Character sets could not contain an opening bracket $[.
2. Fix: Character sets could not contain the characters '[:', e.g. as in '[[:something]' asRegex.

**Comment by Ducasse** [comment]
Thanks a lot. 
BTW ideally we would like to create an FFI binding to the standard perl like library because the problem of our regex is that it is not standard and slow. I do not know if you are interested in this. 
But if you want to give an hand then Pharo could have a nice regex support.

**Comment by bertbarabas** [comment]
I agree that the FFI binding would be ideal although there a few to chose from:

E.g.
http://www.pcre.org/ 
https://github.com/google/re2
https://bitbucket.org/mrabarnett/mrab-regex/src/hg/

So the first step I think would be to pick the best option for Pharo based on features, performance, license etc.

Although maybe this conversation should be separate from this issue…
